### ISSUE TITLE
2.1.7 - GET product - custom_attribute value return not consistent with PUT

### ISSUE THREAD

**Issue by manhao-chen** [magento/magento2#10376]
### Preconditions
currently i'm using the magento 2.1.7 with sample data

### Steps to reproduce
i call the REST api with the following
http://URL/magento-217/rest/V1/products/24-MB04

i get the following response
```
{
	"id": 2,
	"sku": "24-MB04",
	"name": "Strive Shoulder Pack",
	"attribute_set_id": 15,
	"price": 35,
	"status": 1,
	"visibility": 4,
	"type_id": "simple",
	"created_at": "2017-06-11 03:31:25",
	"updated_at": "2017-06-11 03:31:25",
	"extension_attributes": {
		"stock_item": {
			"item_id": 2,
			"product_id": 2,
			"stock_id": 1,
			"qty": 86,
			"is_in_stock": true,
			"is_qty_decimal": false,
			"show_default_notification_message": false,
			"use_config_min_qty": true,
			"min_qty": 0,
			"use_config_min_sale_qty": 1,
			"min_sale_qty": 1,
			"use_config_max_sale_qty": true,
			"max_sale_qty": 10000,
			"use_config_backorders": true,
			"backorders": 0,
			"use_config_notify_stock_qty": true,
			"notify_stock_qty": 1,
			"use_config_qty_increments": true,
			"qty_increments": 0,
			"use_config_enable_qty_inc": true,
			"enable_qty_increments": false,
			"use_config_manage_stock": true,
			"manage_stock": true,
			"low_stock_date": null,
			"is_decimal_divided": false,
			"stock_status_changed_auto": 0
		}
	},
	"product_links": [],
	"options": [],
	"media_gallery_entries": [{
			"id": 2,
			"media_type": "image",
			"label": "Image",
			"position": 1,
			"disabled": false,
			"types": ["image", "small_image", "thumbnail"],
			"file": "\/m\/b\/mb04-black-0.jpg"
		}, {
			"id": 3,
			"media_type": "image",
			"label": "Image",
			"position": 2,
			"disabled": false,
			"types": [],
			"file": "\/m\/b\/mb04-black-0_alt1.jpg"
		}
	],
	"tier_prices": [],
	"custom_attributes": [{
			"attribute_code": "description",
			"value": "<p>Convenience is next to nothing when your day is crammed with action. So whether you're heading to class, gym, or the unbeaten path, make sure you've got your Strive Shoulder Pack stuffed with all your essentials, and extras as well.<\/p>\r\n<ul>\r\n<li>Zippered main compartment.<\/li>\r\n<li>Front zippered pocket.<\/li>\r\n<li>Side mesh pocket.<\/li>\r\n<li>Cell phone pocket on strap.<\/li>\r\n<li>Adjustable shoulder strap and top carry handle.<\/li>\r\n<\/ul>"
		}, {
			"attribute_code": "image",
			"value": "\/m\/b\/mb04-black-0.jpg"
		}, {
			"attribute_code": "small_image",
			"value": "\/m\/b\/mb04-black-0.jpg"
		}, {
			"attribute_code": "thumbnail",
			"value": "\/m\/b\/mb04-black-0.jpg"
		}, {
			"attribute_code": "color",
			"value": "49"
		}, {
			"attribute_code": "news_from_date",
			"value": "2017-06-11 00:00:00"
		}, {
			"attribute_code": "custom_design_from",
			"value": "2017-06-11 00:00:00"
		}, {
			"attribute_code": "category_ids",
			"value": ["3", "7", "4"]
		}, {
			"attribute_code": "options_container",
			"value": "container2"
		}, {
			"attribute_code": "required_options",
			"value": "0"
		}, {
			"attribute_code": "has_options",
			"value": "0"
		}, {
			"attribute_code": "msrp_display_actual_price_type",
			"value": "0"
		}, {
			"attribute_code": "url_key",
			"value": "strive-shoulder-pack"
		}, {
			"attribute_code": "gift_message_available",
			"value": "0"
		}, {
			"attribute_code": "tax_class_id",
			"value": "2"
		}, {
			"attribute_code": "activity",
			"value": "11,21,23,18"
		}, {
			"attribute_code": "style_bags",
			"value": "27,29,30"
		}, {
			"attribute_code": "material",
			"value": "32,33,36,38"
		}, {
			"attribute_code": "strap_bags",
			"value": "61,62,65,66,67"
		}, {
			"attribute_code": "features_bags",
			"value": "72,74,75,78"
		}, {
			"attribute_code": "eco_collection",
			"value": "0"
		}, {
			"attribute_code": "performance_fabric",
			"value": "0"
		}, {
			"attribute_code": "erin_recommends",
			"value": "1"
		}, {
			"attribute_code": "new",
			"value": "0"
		}, {
			"attribute_code": "sale",
			"value": "1"
		}
	]
}
```

### Expected result
PUT request of the GET result should not report back any errors

### Actual result
on my PUT i am just pushing back exactly what was returns from the GET request and the API errors saying
```
{
	"message": "Error occurred during \"custom_attributes\" processing. Attribute \"activity\" has invalid value. Invalid type for value: \"string\". Expected Type: \"string[]\"."
}
```

at this stage i'm gonna assume that all the comma separated customer-attribute values are expecting an array BUT why is it only category_ids is returning as an array but the rest like 

- activity
- style_bags
- material
- strap_bags
- features_bag

all of the above (could be more in other products) coming back as comma separated 
can we please have this standardized so what ever was returned on the GET request could be PUT back as an update of this record

otherwise its extremely difficult to work out which value on return are string, and which ones are string[]

**Comment by magento-engcom-team** [comment]
@manhao-chen thank you for your report.
Please use the issue [reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) to report an issue. Please, edit issue description and include the preconditions, the steps required to reproduce, the actual result, and the expected result.
Please, add the following information:
 - [x] Preconditions: Magento version
 - [x] Steps to reproduce
 - [x] Expected result
 - [x] Actual result

**Comment by manhao-chen** [comment]
@magento-engcom-team updated

**Comment by robsohn** [comment]
The problem doesn't exist in newer version of the code (we tested the issue on 2.3-dev).

It comes down to how Magento\Eav\Model\EavCustomAttributeTypeLocator\SimpleType::getType() works.

For all frontend input types defined as multiselect the method adds `[]` to the type name. Based on the example above we get `string[]`. Of course there is no such type in PHP and hence the exception. 

In newer version of the class (current develop branch) the code appending `[]` has been removed.
I think it can be easily backported by coping the new method to the next 2.1.X release.
